index.js

```diff
@@ -1,130 +1,158 @@
+
+  // extract key paths
+  const timestamp = +new Date();
+  const mainModule = `RunElmMain${timestamp}`;
+  const templatePath = path.resolve(__dirname, 'RunElmMain.elm.template');
+  const templateWithArgsPath = path.resolve(__dirname, 'RunElmMainWithArgs.elm.template');
+  const userModulePath = path.resolve(rawUserModuleFileName);
+  const userModule = path.basename(userModulePath, '.elm');
+  const projectDir = rawProjectDir
+    ? path.resolve(rawProjectDir)
+    : path.resolve(path.dirname(userModulePath));
+  const mainModuleFilename = path.join(projectDir, `${mainModule}.elm`);
+  const outputCompiledFilename = path.join(projectDir, `run-elm-main-${timestamp}.js`);
 
-async function compileElmFile() {
-  sh.cd(outputDir);
+  try {
+    // ensure elm is installed
+    if (!sh.which('elm')) {
+      throw new Error('No elm global binary available. Please install with `npm install -g elm`.');
+    }
 
-  const contents = await compileToString([outputElmFilename], { yes: true });
+    // ensure --output-var is specified adequately
+    if (!outputVarName.match(/^[a-z_]\w*$/)) {
+      throw new Error(`Provided --output-var \`${outputVarName}\` is not a valid variable name in elm.`);
+    }
 
-  return writeFile(outputCompiledFilename, contents);
-}
+    // ensure user moule path is adequate
+    try {
+      const userModuleFileStats = await stat(userModulePath);
+      if (!userModuleFileStats.isFile()) {
+        throw new Error();
+      }
+    } catch (err) {
+      throw new Error(`File '${userModulePath}' does not exist`);
+    }
+    if (!userModulePath.match(/\.elm$/i, '')) {
+      throw new Error(`File '${userModulePath}' should have .elm file extension`);
+    }
 
-function runElmFile(needArgs) {
-  return new Promise((resolve) => {
-    const { worker } = require(outputCompiledFilename)[runModuleName];
-    let app;
+    // ensure project folder is adequate
+    try {
+      const projectDirStats = await stat(projectDir);
+      if (!projectDirStats.isDirectory()) {
+        throw new Error();
+      }
+    } catch (err) {
+      throw new Error(`Provided --project-dir '${rawProjectDir}' is not a directory`);
+    }
+    if (userModulePath.indexOf(`${projectDir}/`) !== 0) {
+      throw new Error(`File \`${userModulePath}\` must be located within --project-dir \`${rawProjectDir}\``);
+    }
 
-    if (needArgs) {
-      const args = process.argv.slice(3);
-      app = worker(args);
-    } else {
-      app = worker();
+    // read user module and determine what template to use
+    let userModuleContents;
+    try {
+      userModuleContents = await readFile(userModulePath, 'utf-8');
+    } catch (err) {
+      throw new Error(`File '${userModulePath}' could not be read`);
+    }
+    const argsRegex = /^output *\w+ *=/;
+    const needArgs = userModuleContents
+      .split('\n')
+      .some(line => argsRegex.test(line.trim()));
+
+    // read main module template
+    await sh.cd(projectDir);
```

Comment:
This should be synchronous already, so we don't need `await`.

Reply:
👌 you're right!